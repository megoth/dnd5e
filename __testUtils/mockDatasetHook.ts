import { mockSolidDatasetFrom, SolidDataset } from "@inrupt/solid-client";
import { SWRResponse } from "swr";

export default function mockDatasetHook(
  mock,
  presets: Partial<SWRResponse<SolidDataset, any>> = {}
) {
  return mock.mockReturnValue({
    data: mockSolidDatasetFrom("http://example.com"),
    error: null,
    ...presets,
  });
}
