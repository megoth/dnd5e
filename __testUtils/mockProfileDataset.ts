import {
  addStringNoLocale,
  addUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf } from "rdf-namespaces";
import { chain } from "../src/utils";

export const aliceWebIdDatasetUrl = "http://example.com/alice";
export const aliceWebIdUrl = `${aliceWebIdDatasetUrl}#me`;
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://example.com/alice.jpg";

export function mockProfileDataset(webId = aliceWebIdUrl) {
  const profile = chain(
    mockThingFrom(webId),
    (t) => addStringNoLocale(t, vcard.fn, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, aliceWebIdUrl)
  );
  return chain(mockSolidDatasetFrom(aliceWebIdDatasetUrl), (d) =>
    setThing(d, profile)
  );
}
