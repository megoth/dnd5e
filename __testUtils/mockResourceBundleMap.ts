import { ResourceBundleModel } from "../src/models/resourceBundle";
import { getBundleKey, ResourceBundleMap } from "../src/models/app";
import { defaultLocale } from "./mockLanguage";
import mockResourceBundle, { defaultBundle } from "./mockResourceBundle";

export default function mockResourceBundleMap(
  overrides: Partial<ResourceBundleModel> = {}
): ResourceBundleMap {
  const label = overrides.label || defaultBundle;
  const locale = overrides.locale || defaultLocale;
  return {
    [getBundleKey(locale, label)]: mockResourceBundle(overrides),
  };
}
