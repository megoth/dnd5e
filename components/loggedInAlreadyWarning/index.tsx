import { useSession } from "@inrupt/solid-ui-react";
import React, { HTMLAttributes } from "react";
import { getThing } from "@inrupt/solid-client";
import WarningMessage from "../warningMessage";
import Translation from "../translation";
import useDataset from "../../src/hooks/useDataset";
import ErrorMessage from "../errorMessage";
import Loading from "../loading";
import getProfileName from "../../src/models/profile";

export const TESTID_LOGGED_IN_ALREADY_WARNING = "logged-in-already-warning";

interface Props extends HTMLAttributes<HTMLDivElement> {}

export default function LoggedInAlreadyWarning({ ...props }: Props) {
  const { session } = useSession();
  const { isLoggedIn, webId } = session.info;
  const { data: profileDataset, error } = useDataset(webId);

  if (!isLoggedIn) {
    return null;
  }

  if (error) {
    return <ErrorMessage error={error} />;
  }

  if (!profileDataset) {
    return <Loading />;
  }

  const profile = getThing(profileDataset, webId);
  const name = getProfileName(profile);

  return (
    isLoggedIn && (
      <div data-testid={TESTID_LOGGED_IN_ALREADY_WARNING} {...props}>
        <WarningMessage>
          <Translation id="LoggedInAlready" vars={{ name }} />
        </WarningMessage>
      </div>
    )
  );
}
