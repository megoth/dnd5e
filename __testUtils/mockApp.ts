import { AppModel, createApp, getBundleKey } from "../src/models/app";
import mockFluentBundle from "./mockFluentBundle";
import { appVocabURL } from "./mockAppIndexDataset";
import mockLanguage, { defaultLocale } from "./mockLanguage";
import mockResourceBundleMap from "./mockResourceBundleMap";
import mockAppIndex from "./mockAppIndex";

export const errorsURL = "https://example.com/global/errors.ttl";
export const faqsURL = "https://example.com/global/faqs.ttl";
export const translationsURL = "https://example.com/global/translations.ttl";
export const localizationsURL =
  "https://example.com/global/translations.en-US.ttl";

export function mockAppPropValue(val, bundleName = "global", locale = "en-US") {
  return bundleName
    ? {
        [getBundleKey(locale, bundleName)]: val,
      }
    : {
        [locale]: val,
      };
}

export default function mockApp(overrides: Partial<AppModel> = {}): AppModel {
  const currentLocale = overrides.currentLocale || defaultLocale;
  return {
    appVocabURL: overrides.appVocabURL || appVocabURL,
    currentLocale,
    fluentBundles: {
      [currentLocale]: mockFluentBundle({}, translationsURL, {
        locale: currentLocale,
      }),
      ...overrides.fluentBundles,
    },
    languages: [mockLanguage(currentLocale), ...(overrides.languages || [])],
    resourceBundles: {
      ...mockResourceBundleMap({
        label: "admin",
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...mockResourceBundleMap({
        locale: currentLocale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
      ...overrides.resourceBundles,
    },
  };
}

export function mockNorwegianApp(overrides: Partial<AppModel> = {}): AppModel {
  const locale = "nb-NO";
  return mockApp({
    currentLocale: locale,
    fluentBundles: {
      [locale]: mockFluentBundle({}, translationsURL, {
        locale,
      }),
    },
    languages: [
      mockLanguage(locale, {
        languageCode: "ðŸ‡³ðŸ‡´",
      }),
    ],
    resourceBundles: {
      ...mockResourceBundleMap({
        locale,
        data: {
          errors: null,
          faqs: null,
          localizations: null,
          translations: null,
        },
      }),
    },
    ...overrides,
  });
}

export function mockEmptyApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, null);
}

export function mockUnloadedApp(): AppModel {
  return createApp(defaultLocale, appVocabURL, mockAppIndex());
}
