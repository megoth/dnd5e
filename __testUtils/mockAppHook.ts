import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
