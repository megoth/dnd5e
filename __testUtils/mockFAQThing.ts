import { mockThingFrom, setUrl } from "@inrupt/solid-client";
import { rdf } from "rdf-namespaces";
import { getAppTerm } from "../src/models/appIndex";
import { appVocabURL } from "./mockAppIndexDataset";
import { chain } from "../src/utils";
import { faqsURL } from "./mockApp";

export const faqId = "faq";
export const faqURL = `${faqsURL}#${faqId}`;
export const faqLabelURL = "https://example.com/#faqLabel";
export const faqDescriptionURL = "https://example.com/#faqDescription";

export default function mockFAQThing(id = faqId) {
  return chain(
    mockThingFrom(`${faqsURL}#${id}`),
    (t) => setUrl(t, rdf.type, getAppTerm("FAQ", { appVocabURL })),
    (t) => setUrl(t, getAppTerm("faqLabel", { appVocabURL }), faqLabelURL),
    (t) =>
      setUrl(
        t,
        getAppTerm("faqDescription", { appVocabURL }),
        faqDescriptionURL
      )
  );
}
