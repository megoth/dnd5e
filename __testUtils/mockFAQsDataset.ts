import {
  mockSolidDatasetFrom,
  setThing,
  SolidDataset,
} from "@inrupt/solid-client";
import { faqsURL } from "./mockApp";
import mockFAQThing from "./mockFAQThing";

export default function mockFAQsDataset(faqs = [mockFAQThing()]) {
  return faqs.reduce<SolidDataset>(
    (memo, faq) => setThing(memo, faq),
    mockSolidDatasetFrom(faqsURL)
  );
}
