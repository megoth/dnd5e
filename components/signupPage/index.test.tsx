import { render } from "@testing-library/react";
import React from "react";
import * as solidUIReactFns from "@inrupt/solid-ui-react";
import { createRouter } from "next/router";
import { RouterContext } from "next/dist/next-server/lib/router-context";
import useApp from "../../src/hooks/useApp";
import mockApp from "../../__testUtils/mockApp";
import mockResourceBundleMap from "../../__testUtils/mockResourceBundleMap";
import mockFAQsDataset from "../../__testUtils/mockFAQsDataset";
import mockFAQThing from "../../__testUtils/mockFAQThing";
import mockAppHook from "../../__testUtils/mockAppHook";
import SignupPage from "./index";
import useDataset from "../../src/hooks/useDataset";
import mockDatasetHook from "../../__testUtils/mockDatasetHook";
import { mockProfileDataset } from "../../__testUtils/mockProfileDataset";
import {
  authenticatedWebId,
  mockAuthenticatedSession,
  mockUnauthenticatedSession,
} from "../../__testUtils/mockSession";
import { TESTID_LOGGED_IN_ALREADY_WARNING } from "../loggedInAlreadyWarning";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

jest.mock("../../src/hooks/useDataset");
const mockedDatasetHook = useDataset as jest.Mock;

const app = mockApp({
  resourceBundles: mockResourceBundleMap({
    data: {
      faqs: mockFAQsDataset([
        mockFAQThing("whatIsIdP"),
        mockFAQThing("whatIsPod"),
      ]),
    },
  }),
});

describe("SignupPage", () => {
  // @ts-ignore
  const router = createRouter("", {}, "", {});

  let mockedSessionHook;

  beforeEach(() => mockAppHook(mockedAppHook, app));
  beforeEach(() => {
    mockedSessionHook = jest
      .spyOn(solidUIReactFns, "useSession")
      .mockReturnValue(mockUnauthenticatedSession());
  });
  beforeEach(() =>
    mockDatasetHook(mockedDatasetHook, {
      data: mockProfileDataset(authenticatedWebId),
    })
  );

  it("renders", () => {
    const { asFragment } = render(
      <RouterContext.Provider value={router}>
        <SignupPage />
      </RouterContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a warning when authenticated", () => {
    mockedSessionHook.mockReturnValue(mockAuthenticatedSession());

    const { asFragment, getByTestId } = render(
      <RouterContext.Provider value={router}>
        <SignupPage />
      </RouterContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTID_LOGGED_IN_ALREADY_WARNING)).toBeDefined();
  });
});
