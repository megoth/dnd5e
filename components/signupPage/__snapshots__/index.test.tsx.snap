// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPage renders 1`] = `
<DocumentFragment>
  <header>
    <a
      class="main-container main-container--content flex focus:outline-none focus:ring-2 focus:ring-red-600"
      href="/"
    >
      <picture>
        <source
          media="(min-width: 640px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-128.png"
        />
        <source
          media="(min-width: 480px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-96.png"
        />
        <img
          alt="[Translation for https://example.com/global/translations.ttl#appLogoDescription does not exist]"
          src="/icons/app/icons8-dungeons-and-dragons-64.png"
        />
      </picture>
      <div
        class="font-serif text-2xl sm:text-3xl md:text-4xl self-center antialiased"
      >
        [Translation for https://example.com/global/translations.ttl#appName does not exist]
      </div>
    </a>
  </header>
  <main
    class="main-container main-container--content my-1"
  >
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <h1>
        [Translation for https://example.com/global/translations.ttl#signupTitle does not exist]
      </h1>
      <p>
        [Translation for https://example.com/global/translations.ttl#signupPitch does not exist]
      </p>
    </div>
    <ul
      class="flex flex-wrap justify-center my-8"
    >
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://signup.pod.inrupt.com/"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/inrupt-ps-logo.png"
          />
          <span
            class="text-xl font-bold"
          >
            Inrupt Pod Spaces
          </span>
        </a>
      </li>
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://solidcommunity.net/register"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/solid-emblem.svg"
          />
          <span
            class="text-xl font-bold"
          >
            solidcommunity.net
          </span>
        </a>
      </li>
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://inrupt.net/register"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/solid-emblem.svg"
          />
          <span
            class="text-xl font-bold"
          >
            inrupt.net
          </span>
        </a>
      </li>
    </ul>
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <p
        class="text-center"
      >
        <a
          href="/login"
        >
          [Translation for https://example.com/global/translations.ttl#logInPrompt does not exist]
        </a>
      </p>
    </div>
    <aside
      class="my-16"
    >
      <div
        class="content leading-normal max-w-prose mx-auto content content--small"
      >
        <h2
          data-testid="faq-label"
        >
          [Translation for https://example.com/#faqLabel does not exist]
        </h2>
        <div
          data-testid="faq-description"
        >
          <p>
            [Translation for https://example.com/#faqDescription does not exist]
          </p>
        </div>
      </div>
      <div
        class="content leading-normal max-w-prose mx-auto content content--small"
      >
        <h2
          data-testid="faq-label"
        >
          [Translation for https://example.com/#faqLabel does not exist]
        </h2>
        <div
          data-testid="faq-description"
        >
          <p>
            [Translation for https://example.com/#faqDescription does not exist]
          </p>
        </div>
      </div>
    </aside>
  </main>
  <footer
    class="main-container main-container--content mt-2 text-center sm:text-left xs:flex xs:space-x-4 items-center justify-center sm:justify-start"
  >
    <div
      class="md:flex-grow pb-2 flex space-x-1 justify-center sm:justify-start"
    >
      <a
        class="link"
        href="/about"
      >
        [Translation for https://example.com/global/translations.ttl#aboutTitle does not exist]
      </a>
      <a
        class="link"
        href="/faq"
      >
        [Translation for https://example.com/global/translations.ttl#faqShort does not exist]
      </a>
    </div>
    <div
      class="pb-2 flex xs:space-x-2 space-y-2 xs:space-y-0 flex-col text-center xs:flex-row xs:justify-center"
    >
      <button
        class="button"
        data-testid="dark-mode-selector-button"
        type="button"
      >
        <span>
          <span
            class="text-black text-opacity-30"
          >
            ðŸ’¡
          </span>
          Â [Translation for https://example.com/global/translations.ttl#darkModeTurnOn does not exist]
        </span>
      </button>
    </div>
  </footer>
</DocumentFragment>
`;

exports[`SignupPage renders a warning when authenticated 1`] = `
<DocumentFragment>
  <header>
    <a
      class="main-container main-container--content flex focus:outline-none focus:ring-2 focus:ring-red-600"
      href="/"
    >
      <picture>
        <source
          media="(min-width: 640px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-128.png"
        />
        <source
          media="(min-width: 480px)"
          srcset="/icons/app/icons8-dungeons-and-dragons-96.png"
        />
        <img
          alt="[Translation for https://example.com/global/translations.ttl#appLogoDescription does not exist]"
          src="/icons/app/icons8-dungeons-and-dragons-64.png"
        />
      </picture>
      <div
        class="font-serif text-2xl sm:text-3xl md:text-4xl self-center antialiased"
      >
        [Translation for https://example.com/global/translations.ttl#appName does not exist]
      </div>
    </a>
  </header>
  <main
    class="main-container main-container--content my-1"
  >
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <h1>
        [Translation for https://example.com/global/translations.ttl#signupTitle does not exist]
      </h1>
      <div
        data-testid="logged-in-already-warning"
      >
        <div
          class="bg-yellow-200 text-yellow-900 border-yellow-400 border px-2 py-1 rounded-sm mb-2"
        >
          [Translation for https://example.com/global/translations.ttl#LoggedInAlready does not exist]
        </div>
      </div>
      <p>
        [Translation for https://example.com/global/translations.ttl#signupPitch does not exist]
      </p>
    </div>
    <ul
      class="flex flex-wrap justify-center my-8"
    >
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://signup.pod.inrupt.com/"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/inrupt-ps-logo.png"
          />
          <span
            class="text-xl font-bold"
          >
            Inrupt Pod Spaces
          </span>
        </a>
      </li>
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://solidcommunity.net/register"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/solid-emblem.svg"
          />
          <span
            class="text-xl font-bold"
          >
            solidcommunity.net
          </span>
        </a>
      </li>
      <li
        class="text-center m-2"
      >
        <a
          class="block bg-purple-100 p-4 rounded-xl focus:outline-none focus:ring-2 focus:ring-purple-600 hover:shadow-inner hover:bg-purple-300"
          href="https://inrupt.net/register"
        >
          <img
            alt="[Translation for https://example.com/global/translations.ttl#serviceLogo does not exist]"
            class="mx-auto max-h-20"
            src="/logos/solid-emblem.svg"
          />
          <span
            class="text-xl font-bold"
          >
            inrupt.net
          </span>
        </a>
      </li>
    </ul>
    <div
      class="content leading-normal max-w-prose mx-auto"
    >
      <p
        class="text-center"
      >
        <a
          href="/login"
        >
          [Translation for https://example.com/global/translations.ttl#logInPrompt does not exist]
        </a>
      </p>
    </div>
    <aside
      class="my-16"
    >
      <div
        class="content leading-normal max-w-prose mx-auto content content--small"
      >
        <h2
          data-testid="faq-label"
        >
          [Translation for https://example.com/#faqLabel does not exist]
        </h2>
        <div
          data-testid="faq-description"
        >
          <p>
            [Translation for https://example.com/#faqDescription does not exist]
          </p>
        </div>
      </div>
      <div
        class="content leading-normal max-w-prose mx-auto content content--small"
      >
        <h2
          data-testid="faq-label"
        >
          [Translation for https://example.com/#faqLabel does not exist]
        </h2>
        <div
          data-testid="faq-description"
        >
          <p>
            [Translation for https://example.com/#faqDescription does not exist]
          </p>
        </div>
      </div>
    </aside>
  </main>
  <footer
    class="main-container main-container--content mt-2 text-center sm:text-left xs:flex xs:space-x-4 items-center justify-center sm:justify-start"
  >
    <div
      class="md:flex-grow pb-2 flex space-x-1 justify-center sm:justify-start"
    >
      <a
        class="link"
        href="/about"
      >
        [Translation for https://example.com/global/translations.ttl#aboutTitle does not exist]
      </a>
      <a
        class="link"
        href="/faq"
      >
        [Translation for https://example.com/global/translations.ttl#faqShort does not exist]
      </a>
    </div>
    <div
      class="pb-2 flex xs:space-x-2 space-y-2 xs:space-y-0 flex-col text-center xs:flex-row xs:justify-center"
    >
      <button
        class="button"
        data-testid="dark-mode-selector-button"
        type="button"
      >
        <span>
          <span
            class="text-black text-opacity-30"
          >
            ðŸ’¡
          </span>
          Â [Translation for https://example.com/global/translations.ttl#darkModeTurnOn does not exist]
        </span>
      </button>
    </div>
  </footer>
</DocumentFragment>
`;
